{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Written for Jupyter Notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Import Requirements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cairo\n",
    "import math\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_fa(path_to_fasta, path_to_outputfile):\n",
    "    '''Condenses seq info to a single line. Outputs a copy of input file, but with all seq info on one line. Each read consists of only 2 lines (header and seq). Output file does not necessarily have to be already created'''\n",
    "    \n",
    "    # Open file and loop through file (one line at a time)\n",
    "    with open(path_to_fasta, \"r\") as in_fh, open(path_to_outputfile, \"w\") as temp_fh:\n",
    "\n",
    "        file_line1 = 'true' # first line in entire file\n",
    "        \n",
    "        for line in in_fh:\n",
    "            \n",
    "            # Write very first line in file to output\n",
    "            if file_line1 == 'true':\n",
    "                temp_fh.write(line)\n",
    "                # update that no longer looking at first line in entire file\n",
    "                file_line1 = 'false'\n",
    "            \n",
    "            # Looking at any line that isn't first line in file\n",
    "            else:\n",
    "                # If seq line --> strip whitespace and write to output\n",
    "                if line[0] != \">\":\n",
    "                    seq_line = line.strip()\n",
    "                    temp_fh.write(seq_line)\n",
    "\n",
    "                # If header line --> add newline char and write to output \n",
    "                if line[0] == \">\":\n",
    "                    temp_fh.write(\"\\n\" + line) \n",
    "                    \n",
    "        # Proper format: end of file - has last seq line end with \\n \n",
    "        temp_fh.write(\"\\n\")\n",
    "\n",
    "    return \"Fasta reformatting completed. Temp file located at:\", temp_path\n",
    "\n",
    "## Test above function works ##\n",
    "# fa_path=\"../data/testfile_INSR.fasta\"\n",
    "# temp_path=\"../data/temp.fa\"\n",
    "# format_fa(fa_path, temp_path) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_nt_type(nt):\n",
    "    '''Given a single nt, outputs string if exon (capital letter) or intron (lowercase letter)'''\n",
    "    if nt.isupper() == True:\n",
    "        current_type = 'exon'\n",
    "        \n",
    "    if nt.islower() == True:\n",
    "        current_type = 'intron'\n",
    "        \n",
    "    return str(current_type)\n",
    "\n",
    "## Test above function works #\n",
    "# get_nt_type(\"z\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ambig_found(target_motif_list):\n",
    "    '''Input list of motif(s) that want to find in a read seq later on.\n",
    "    Checks if only ATCG or atcg in motif seq. Returns False if only ATCG \n",
    "    or 'true' if any ambiguous characters found \n",
    "    (e.g. Y, etc).'''\n",
    "    import re\n",
    "    # Code below will exit function at first sight of \n",
    "        # any motif with ambi char\n",
    "        # even if not finished looping through motif list\n",
    "        \n",
    "    i = 0\n",
    "    for motif in target_motif_list:\n",
    "        \n",
    "        match = re.search(r\"[^ATCGatcg]\", motif) # match = match object in re\n",
    "\n",
    "        # If motif has ambig character(s)\n",
    "        if match:\n",
    "            return 'true'\n",
    "\n",
    "        # If only AGTC\n",
    "        else:\n",
    "            # look at next motif in list for ambig char\n",
    "            i+=1\n",
    "    return 'false'\n",
    "\n",
    "\n",
    "## Test above function works ##\n",
    "# test_list = ['agag', 'aaaq']\n",
    "# ambig_found(test_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "# must already created a dict containing ambig and their meanings = called ambig_dna_dict\n",
    "\n",
    "def replace_ambig_char(motif_list):\n",
    "    '''Input motif list and return mod_motif_list with these converted ambig characters, checks if ambig char present,\n",
    "    outputs motif without ambig char as per regex.\n",
    "    Motif seq can be capital or lowercase\n",
    "    motif = yaa outputs to motif = [AT]aa. Assumes you have\n",
    "    already created ambig_dna_dict where keys = ambig char\n",
    "    and values = all possible nt that ambig char could be'''\n",
    "    import re\n",
    "    \n",
    "    mod_motif_list = []\n",
    "\n",
    "    i = 0\n",
    "    for motif in motif_list:\n",
    "        for key in ambig_dna_dict:\n",
    "            if key in motif:\n",
    "            # flags = allows to work for lower or capital letters, e.g. input yaa, sees key=y value=A or T, output [AT]aa\n",
    "                meaning = ambig_dna_dict[key]\n",
    "                \n",
    "                mod_motif = re.sub(key, meaning, motif, flags = re.IGNORECASE)\n",
    "                motif = mod_motif\n",
    "                \n",
    "                \n",
    "\n",
    "        mod_motif_list.append(motif)\n",
    "        i+=1\n",
    "\n",
    "    return mod_motif_list\n",
    "\n",
    "## Test above function works ##\n",
    "# test_list = ['aaa', 'yat', 'ggg']\n",
    "# replace_ambig_char(test_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mot_coords(motif, nt_seq):\n",
    "    '''scan if a motif in seq, then return start/stop coords, draw box, continue scanning for motif'''\n",
    "    import re\n",
    "    \n",
    "    # list to store all start/stop coords for the given motif\n",
    "    coords = []\n",
    "    \n",
    "    # look for motif in nt_seq, flags makes it work for both capital and lowercase letters\n",
    "    match = re.finditer(motif, nt_seq, flags=re.IGNORECASE)\n",
    "        \n",
    "    if match:\n",
    "        i=0\n",
    "        for item in match:\n",
    "            start = item.start()+1\n",
    "            stop = item.end()\n",
    "            coords.append(start)\n",
    "            coords.append(stop)\n",
    "            i+=1\n",
    "        return coords\n",
    "    \n",
    "    else:\n",
    "        return # returns empty list if motif not found in seq\n",
    "    \n",
    "    \n",
    "## Test if above function works ##\n",
    "# test_motif = 'gggg'\n",
    "# test_nt_seq = 'atataggggtatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatataggggtatatatccccatatatatatatatatatatatatatatatAAAAACATACCCCGGCACTGGTGTCGAGGGGATAGatatataatatatatatatatatatatatatatatatatatatatatatatatatatatatatatccccatatatatatatatatatatatatatatatatatatatatatatatatataggggtatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatatccttccttccttctttctttttttccttctttctttccttctttcttttc'\n",
    "# mot_coords(test_motif, test_nt_seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Code Body"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "################# Arg parse #################\n",
    "# user needs to tell how many total reads in file, num_rds\n",
    "num_rds = 2\n",
    "\n",
    "\n",
    "\n",
    "################# Obj set up or comment out for arg parse #################\n",
    "fa_path=\"../data/testfile_INSR.fasta\"\n",
    "# fa_path=\"../data/testfile_INSR.1read.fasta\"\n",
    "temp_path=\"../data/temp.fa\"\n",
    "# temp_path=\"../data/temp.1read.fa\"\n",
    "motif_path=\"../data/testfile_motifs.txt\" # contains y\n",
    "# motif_path=\"../data/motifs.txt\" # contiains y\n",
    "\n",
    "\n",
    "################# Reformat fa: all seq info on 1 line  #################        \n",
    "# Create temp file that is copy of original fa\n",
    "#but reformatted for prot seq on one line only \n",
    "# Will use temp file as fasta for all downstream work\n",
    "format_fa(fa_path, temp_path) \n",
    "\n",
    "\n",
    "################# Save motifs of interest in motif_list #################\n",
    "with open(motif_path, \"r\") as motif_fh:\n",
    "    # Initialize list\n",
    "    motif_list = []\n",
    "    # Add motifs to list\n",
    "    for line in motif_fh:\n",
    "        line = line.strip()\n",
    "        motif_list.append(line)\n",
    "\n",
    "\n",
    "################# Create_ambig_dict,  key = nt code, value = tuple of nts it represents #################\n",
    "# ambig_dna_dict contains for ambiguous nt meanings, \n",
    "    #e.g. Y represents a nt that can be a cytosine or thymine.\n",
    "    #Only works for Y, R, W, S, K, M, D, , H, B, X, and N.\n",
    "    #Will populate dict containing lower and uppercase versions\n",
    "    \n",
    "# Create dict ambig_dna, \n",
    "ambig_dna_dict = {}\n",
    "nt_key = [\"U\", \"u\", \"Y\", \"y\", \"R\", \"r\", \"W\", \"w\", \"S\", \"s\", \"K\", \"k\", \\\n",
    "            \"M\", \"m\", \"D\", \"d\", \"V\", \"v\", \"H\", \"h\", \"B\", \"b\", \\\n",
    "            \"X\", \"x\", \"N\", \"n\"]\n",
    "repre = ['[T]', '[t]', '[CT]', '[ct]', '[AG]', '[ag]', '[AT]', '[at]', '[GC]', '[gc]', '[TG]', '[tg]', \\\n",
    "         '[CA]', '[ca]', '[AGT]', '[agt]', '[ACG]', '[acg]', '[ACT]', '[act]', '[CGT]', '[cgt]', \\\n",
    "         '[ATCG]', '[atcg]', '[ATCG]', '[atcg]']\n",
    "# Populate ambig_dna_dict\n",
    "i = 0\n",
    "for symbol in nt_key:\n",
    "    meaning = repre[i]\n",
    "    # Populate dict\n",
    "    ambig_dna_dict[symbol] = meaning\n",
    "    i+=1 \n",
    "# print(ambig_dna_dict)\n",
    "\n",
    "\n",
    "################# Set up for using pycairo #################\n",
    "# Outline for all remaining code:\n",
    "#Notes: 1)loop through one read and plot exons/introns, 2)loop thorugh same read and plot motifs, 3)move to next read and repeat\n",
    "\n",
    "# Set up 'canvas' where drawing image\n",
    "width, height = 800, 500\n",
    "# Create coordaintes where to display image \n",
    "    # produces SVG file\n",
    "svgfile = \"../output/testfile_figure.svg\"\n",
    "surface = cairo.SVGSurface(svgfile, width, height)\n",
    "# Create create the coordinates where I will be drawing on\n",
    "context = cairo.Context(surface)\n",
    "# Width of line\n",
    "context.set_line_width(1)\n",
    "# All set to draw image downstream\n",
    "\n",
    "################# Read file once: Draw intron line #################\n",
    "\n",
    "with open(temp_path, 'r') as mod_fa:\n",
    "    # set starting pt\n",
    "    x_init = 5\n",
    "    y_init = 160 \n",
    "    rd_counter = 1\n",
    "    \n",
    "    while True:\n",
    "        header = mod_fa.readline().strip()\n",
    "        seq = mod_fa.readline().strip()\n",
    "        if not header: # if end of file\n",
    "            break  \n",
    "        \n",
    "\n",
    "        if seq:\n",
    "            ## 1. Set up Canvas ##\n",
    "            # Scale canvas, width = full width of canvas\n",
    "            draw_width = width - 10\n",
    "            # Determine length of seq for a given read and incorporate into scaling\n",
    "            seq_length = len(seq)\n",
    "            adj_seq_w = draw_width / seq_length # how many canvas units represent 1 nt\n",
    "            \n",
    "            ## 2. Draw horizontal line that is as long as total nts in seq (then scaled) ##\n",
    "            # how far down to start, arbitrarily picked 4\n",
    "            y = y_init + (40 *(rd_counter -1))\n",
    "#             y = y_init + (rd_counter * 4)\n",
    "            # Move to starting spot of line\n",
    "            context.move_to(x_init, y)\n",
    "            # Move to ending spot of line (aka how forare right to move)\n",
    "            x = (x_init + seq_length) * adj_seq_w # x = (total nts) * canvas_units_per_1_nt\n",
    "            context.line_to(x, y)\n",
    "            # Draw it\n",
    "            context.stroke()\n",
    "            # Update rd counter\n",
    "            rd_counter +=1\n",
    "            \n",
    "################# Read file Second Time: Draw exon boxes #################\n",
    "with open(temp_path, 'r') as mod_fa:\n",
    "    # set starting pt\n",
    "    x_init = 5\n",
    "    y_init = 160\n",
    "    rd_counter = 1\n",
    "    font_x_init = 5\n",
    "    font_y_init = 130\n",
    "    \n",
    "    while True:\n",
    "        header = mod_fa.readline().strip()\n",
    "        seq = mod_fa.readline().strip()\n",
    "        if not header: # if end of file\n",
    "            break\n",
    "        \n",
    "        if header:\n",
    "            ## Write out legend ##\n",
    "            context.select_font_face(\"Times New Roman\", cairo.FONT_SLANT_NORMAL, cairo.FONT_WEIGHT_NORMAL)\n",
    "            context.set_font_size(16)\n",
    "            x_leg = 5\n",
    "            y_leg = 25\n",
    "            context.move_to(x_leg, y_leg)\n",
    "            context.show_text('Legend') #add legend name\n",
    "            # Exon legend\n",
    "            context.move_to(x_leg, y_leg+20)\n",
    "            context.rectangle(x_leg, y_leg+20, 5, 20)\n",
    "            context.fill()\n",
    "            context.move_to(x_leg+10, y_leg+35)\n",
    "            context.show_text('Exon')\n",
    "            # Intron legend\n",
    "            context.move_to(x_leg+80, y_leg+30)\n",
    "            context.line_to(x_leg+110, y_leg+30)\n",
    "            context.stroke()\n",
    "            context.move_to(x_leg+120, y_leg+35)\n",
    "            context.show_text('Intron')\n",
    "            # Motif legend\n",
    "            col = 1\n",
    "            move = 0\n",
    "            for motif in motif_list:\n",
    "                # Draw colored box\n",
    "                context.move_to(x_leg+move, y_leg+55)\n",
    "                context.rectangle(x_leg+move, y_leg+55, 5, 20)\n",
    "                context.set_source_rgb(0.45*col, 0.25*col, 0.15)\n",
    "                context.fill()\n",
    "                # write text\n",
    "                context.move_to(x_leg+10+move, y_leg+70)\n",
    "                context.set_source_rgb(0,0,0)\n",
    "                context.show_text(motif)\n",
    "                col +=0.75\n",
    "                move+=100\n",
    "\n",
    "            ## Type out header info to output ##\n",
    "            context.set_source_rgb(0,0,0)\n",
    "            context.select_font_face(\"Times New Roman\", cairo.FONT_SLANT_NORMAL, cairo.FONT_WEIGHT_NORMAL)\n",
    "            context.set_font_size(13)\n",
    "            font_y = font_y_init + (40 *(rd_counter -1) + 15)\n",
    "            context.move_to(font_x_init, font_y)\n",
    "            context.show_text(header)\n",
    "        \n",
    "        if seq:\n",
    "            ## 1. Set up Canvas ##\n",
    "            # Scale canvas, width = full width of canvas\n",
    "            draw_width = width - 10\n",
    "            # Determine length of seq for a given read and incorporate into scaling\n",
    "            seq_length = len(seq)\n",
    "            adj_seq_w = draw_width / seq_length # how many canvas units represent 1 nt\n",
    "#             print('rd_counter is', rd_counter)\n",
    "            \n",
    "    \n",
    "            ## 2. Draw exons ##\n",
    "            i = 0\n",
    "            start = 'no exon yet'\n",
    "            stop = 'null'\n",
    "            for nt in seq:\n",
    "                if start == 'no exon yet':\n",
    "                    # If intron, then move on to next pos\n",
    "                    if get_nt_type(nt) == 'intron':\n",
    "                        i+=1\n",
    "                    # If exon\n",
    "                    if get_nt_type(nt) == 'exon':\n",
    "                        start = i\n",
    "                        i+=1\n",
    "                elif stop == 'null':\n",
    "                    # if still see exon\n",
    "                    if get_nt_type(nt) == 'exon':\n",
    "                        i+=1\n",
    "                    # if now see introns\n",
    "                    if get_nt_type(nt) == 'intron':\n",
    "                        stop = i\n",
    "#                         print(start, stop)\n",
    "                        \n",
    "                        # Now have start and stop coordinates of exon..let's draw\n",
    "                        context.move_to(x_init, y_init)\n",
    "                        # Draw rect cords. \n",
    "                        #x0 --> x coord of top left rect corner\n",
    "                        #y0 --> y coord of top left rect corner\n",
    "                        #x1 --> rectangle lenght (how far right to go)\n",
    "                        #y1 --> rectangle height (how far down to go)   \n",
    "#                         print('exon start and stop coords:', start, stop)\n",
    "                        x0 = start * adj_seq_w # multip by canvas units per 1 nt\n",
    "                        x1 = (stop-start) * adj_seq_w # multip by canvas units per 1 nt\n",
    "                        y0 = (y_init + (40 *(rd_counter -1))) - 10\n",
    "                        y1 = 20\n",
    "                        context.rectangle(x0, y0, x1, y1)\n",
    "                        context.fill()\n",
    "                        # Reset for next nt in seq\n",
    "                        start = 'no exon yet'\n",
    "                        stop = 'null'\n",
    "                        i+=1\n",
    "                  \n",
    "                \n",
    "            # Update things before moving to next read in fadd\n",
    "            rd_counter +=1\n",
    "                           \n",
    "                \n",
    "################# Read file Third time: Draw motif boxes #################\n",
    "\n",
    "with open(temp_path, 'r') as mod_fa:\n",
    "    # set starting pt\n",
    "    x_init = 5\n",
    "    y_init = 160\n",
    "    rd_counter = 1\n",
    "    \n",
    "    while True:\n",
    "        header = mod_fa.readline().strip()\n",
    "        seq = mod_fa.readline().strip()\n",
    "        if not header: # if end of file\n",
    "            break\n",
    "        \n",
    "        if seq:\n",
    "            \n",
    "            ## 4. Replace any ambig char that exist in motifs of interest ##\n",
    "            if ambig_found(motif_list) == 'true':\n",
    "                # Create mod_mofit_list that has ambigs as regex, y --> [ct]\n",
    "                motif_list = replace_ambig_char(motif_list)\n",
    "#                 print(motif_list)\n",
    "                \n",
    "                \n",
    "                \n",
    "            ## 5. Draw motif ##\n",
    "            col=1\n",
    "            for motif in motif_list:\n",
    "#                 create function: scan if a motif in seq, then return start/stop coords, draw box, continue scanning for motif\n",
    "                start_stop = mot_coords(motif, seq)\n",
    "#                 print(start_stop)\n",
    "                \n",
    "                \n",
    "                \n",
    "                even = 0\n",
    "                odd = even + 1\n",
    "        \n",
    "        \n",
    "                for cood in start_stop:\n",
    "                    if odd < len(start_stop):\n",
    "#                         print('even and odd is', even, odd)\n",
    "                        start = start_stop[even]\n",
    "                        stop = start_stop[odd]\n",
    "#                         print(start, stop)\n",
    "\n",
    "                        x0 = start * adj_seq_w # multip by canvas units per 1 nt\n",
    "                        x1 = (stop-start) * adj_seq_w # multip by canvas units per 1 nt\n",
    "                        y0 = (y_init + (40 *(rd_counter -1))) - 10\n",
    "                        y1 = 20\n",
    "                        context.rectangle(x0, y0, x1, y1)\n",
    "                        context.set_source_rgb(0.45*col, 0.25*col, 0.15)\n",
    "                        context.fill()\n",
    "\n",
    "                        #update \n",
    "                        even +=2\n",
    "                        odd+=2\n",
    "                col+=0.75\n",
    "                        \n",
    "            rd_counter+=1\n",
    "\n",
    "\n",
    "                \n",
    "\n",
    "\n",
    "                \n",
    "# Image complete\n",
    "surface.finish()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Written for Command Line \n",
    "\n",
    "#### Contains info for arg parse\n",
    "\n",
    "Note unable to get pycairo to run on local computer. Instead ran this code on Talapas and `scp` the output files to my local computer\n",
    "\n",
    "##### Code is saved as draw_motifs.py in this same dir"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
